File: BOJ(Java 공부용)/9012.java
import java.io.*;
import java.util.ArrayDeque;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        int n = Integer.parseInt(input[0]);
        for(int i=0;i<n;i++) {
            ArrayDeque<Character> st = new ArrayDeque<>();
            boolean check = false;
            String[] s = br.readLine().split(" ");
            for(int j=0;j<s[0].length();j++) {
                if(s[0].charAt(j) == '(') {
                    st.push(s[0].charAt(j));
                }else{
                    if(st.isEmpty()) {
                        check = true;
                        break;
                    }
                    st.pop();
                }
            }
            if(check || !st.isEmpty()) bw.write("NO\n");
            else bw.write("YES\n");
        }
        bw.flush();
    }
}

File: BOJ(Kotlin)/9012.kt
import com.sun.org.apache.xpath.internal.operations.Bool
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

fun main(args : Array<String>) {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    var st = StringTokenizer(br.readLine())
    val n = st.nextToken().toInt()

    for (i in 0 until n) {
        val dq : Deque<Char> = ArrayDeque()
        val s = br.readLine()
        var check = false
        for (j in s.indices) { // s.indices : Returns the range of valid character indices for this char sequence
            if (s[j] == '(') {
                dq.push(s[j])
            }
            else {
                if (dq.isEmpty()) {
                    check = true
                    break
                }
                dq.pop()
            }
        }

        if (check || !dq.isEmpty()) bw.write("NO\n")
        else bw.write("YES\n")
    }

    bw.flush()
    bw.close()
}